File: frontend/src/lib/gst/stateCodes.ts
export interface GSTState {
  code: string;
  name: string;
  isUnionTerritory: boolean;
}

export const GST_STATE_CODES: Record<string, GSTState> = {
  "01": { code: "01", name: "Jammu and Kashmir", isUnionTerritory: false },
  "02": { code: "02", name: "Himachal Pradesh", isUnionTerritory: false },
  "03": { code: "03", name: "Punjab", isUnionTerritory: false },
  "04": { code: "04", name: "Chandigarh", isUnionTerritory: true },
  "05": { code: "05", name: "Uttarakhand", isUnionTerritory: false },
  "06": { code: "06", name: "Haryana", isUnionTerritory: false },
  "07": { code: "07", name: "Delhi", isUnionTerritory: true },
  "08": { code: "08", name: "Rajasthan", isUnionTerritory: false },
  "09": { code: "09", name: "Uttar Pradesh", isUnionTerritory: false },
  "10": { code: "10", name: "Bihar", isUnionTerritory: false },
  "11": { code: "11", name: "Sikkim", isUnionTerritory: false },
  "12": { code: "12", name: "Arunachal Pradesh", isUnionTerritory: false },
  "13": { code: "13", name: "Nagaland", isUnionTerritory: false },
  "14": { code: "14", name: "Manipur", isUnionTerritory: false },
  "15": { code: "15", name: "Mizoram", isUnionTerritory: false },
  "16": { code: "16", name: "Tripura", isUnionTerritory: false },
  "17": { code: "17", name: "Meghalaya", isUnionTerritory: false },
  "18": { code: "18", name: "Assam", isUnionTerritory: false },
  "19": { code: "19", name: "West Bengal", isUnionTerritory: false },
  "20": { code: "20", name: "Jharkhand", isUnionTerritory: false },
  "21": { code: "21", name: "Odisha", isUnionTerritory: false },
  "22": { code: "22", name: "Chhattisgarh", isUnionTerritory: false },
  "23": { code: "23", name: "Madhya Pradesh", isUnionTerritory: false },
  "24": { code: "24", name: "Gujarat", isUnionTerritory: false },
  "25": { code: "25", name: "Daman and Diu", isUnionTerritory: true },
  "26": { code: "26", name: "Dadra and Nagar Haveli and Daman and Diu", isUnionTerritory: true },
  "27": { code: "27", name: "Maharashtra", isUnionTerritory: false },
  "29": { code: "29", name: "Karnataka", isUnionTerritory: false },
  "30": { code: "30", name: "Goa", isUnionTerritory: false },
  "31": { code: "31", name: "Lakshadweep", isUnionTerritory: true },
  "32": { code: "32", name: "Kerala", isUnionTerritory: false },
  "33": { code: "33", name: "Tamil Nadu", isUnionTerritory: false },
  "34": { code: "34", name: "Puducherry", isUnionTerritory: true },
  "35": { code: "35", name: "Andaman and Nicobar Islands", isUnionTerritory: true },
  "36": { code: "36", name: "Telangana", isUnionTerritory: false },
  "37": { code: "37", name: "Andhra Pradesh", isUnionTerritory: false },
  "38": { code: "38", name: "Ladakh", isUnionTerritory: true },
  "97": { code: "97", name: "Other Territory", isUnionTerritory: true },
  "99": { code: "99", name: "Centre Jurisdiction", isUnionTerritory: true },
};

export function getStateByCode(code: string): GSTState | null {
  return GST_STATE_CODES[code] ?? null;
}

export function getAllStates(): GSTState[] {
  return Object.values(GST_STATE_CODES).sort((a, b) =>
    a.code.localeCompare(b.code)
  );
}

File: backend/app/core/gst/state_codes.py
GST_STATE_CODES = {
    "01": ("Jammu and Kashmir", False),
    "02": ("Himachal Pradesh", False),
    "03": ("Punjab", False),
    "04": ("Chandigarh", True),
    "05": ("Uttarakhand", False),
    "06": ("Haryana", False),
    "07": ("Delhi", True),
    "08": ("Rajasthan", False),
    "09": ("Uttar Pradesh", False),
    "10": ("Bihar", False),
    "11": ("Sikkim", False),
    "12": ("Arunachal Pradesh", False),
    "13": ("Nagaland", False),
    "14": ("Manipur", False),
    "15": ("Mizoram", False),
    "16": ("Tripura", False),
    "17": ("Meghalaya", False),
    "18": ("Assam", False),
    "19": ("West Bengal", False),
    "20": ("Jharkhand", False),
    "21": ("Odisha", False),
    "22": ("Chhattisgarh", False),
    "23": ("Madhya Pradesh", False),
    "24": ("Gujarat", False),
    "25": ("Daman and Diu", True),
    "26": ("Dadra and Nagar Haveli and Daman and Diu", True),
    "27": ("Maharashtra", False),
    "29": ("Karnataka", False),
    "30": ("Goa", False),
    "31": ("Lakshadweep", True),
    "32": ("Kerala", False),
    "33": ("Tamil Nadu", False),
    "34": ("Puducherry", True),
    "35": ("Andaman and Nicobar Islands", True),
    "36": ("Telangana", False),
    "37": ("Andhra Pradesh", False),
    "38": ("Ladakh", True),
    "97": ("Other Territory", True),
    "99": ("Centre Jurisdiction", True),
}

class GSTStateMaster:
    @staticmethod
    def get_state(code: str) -> dict | None:
        state_data = GST_STATE_CODES.get(code)
        if not state_data:
            return None
        return {
            "code": code,
            "name": state_data[0],
            "is_union_territory": state_data[1]
        }

    @staticmethod
    def get_state_name(code: str) -> str | None:
        state_data = GST_STATE_CODES.get(code)
        return state_data[0] if state_data else None

    @staticmethod
    def is_valid_state_code(code: str) -> bool:
        return code in GST_STATE_CODES

    @staticmethod
    def all_states() -> list[dict]:
        return [
            {
                "code": code,
                "name": data[0],
                "is_union_territory": data[1]
            }
            for code, data in sorted(GST_STATE_CODES.items())
        ]

    @staticmethod
    def get_by_name(name: str) -> dict | None:
        name_lower = name.lower()
        for code, data in GST_STATE_CODES.items():
            if data[0].lower() == name_lower:
                return {
                    "code": code,
                    "name": data[0],
                    "is_union_territory": data[1]
                }
        return None
